wwwroot/assets/scripts/legend.js

```diff
@@ -0,0 +1,111 @@
+export default function () {
+  let legend = {
+    //* Data structure: 
+    //* Indicators {
+    //*   IndicatorID1: { 
+    //*      'indicator_name': '',
+    //*      Year1: {'reduced': false,
+    //*              'values': [ , ]
+    //*             },
+    //*      Year2: { }
+    //*   }
+    //* }
+
+    indicators: {},
+
+    setValue(indicatorId, indicatorName, year, value) {
+      if (this.indicators[indicatorId]) {
+        if (this.indicators[indicatorId][year]) {
+          // reduced = true means we already processed all data for this year and indicator
+          if (this.indicators[indicatorId][year].reduced === false) {
+            // add value to existing array for year and indicatorID
+            let arr = this.indicators[indicatorId][year]['values'];
+            arr.push(value);
+            this.indicators[indicatorId][year]['values'] = arr;
+          }
+        }
+        else {
+          // create new year property for indicatorID with single elt array
+          this.indicators[indicatorId][year] = {'reduced': false, 'values':[value]};
+        }
+      } else {
+        // create new indicator property
+        let yearObj = {};
+        yearObj[year] = {'reduced': false, 'values':[value]};
+        this.indicators[indicatorId] = yearObj;
+        this.indicators[indicatorId]['indicator_name'] = indicatorName;
+      }
+    },
+
+    getValues(indicatorId, year) {
+      if (this.indicators[indicatorId]) {
+        if (this.indicators[indicatorId][year]) {
+          return this.indicators[indicatorId][year].values;
+        }
+      }
+      return null;
+    },
+
+    getYears(indicatorId) {
+      let yearArray = [];
+      // only get years relevant to active indicator
+      for (let yearVal in this.indicators[indicatorId]) {
+        let year = this.indicators[indicatorId][yearVal];
+        if (typeof year.reduced === 'boolean') {
+          //if (!yearArray.includes(yearVal)) yearArray.push(yearVal);
+          yearArray.push(yearVal);
+        }
+      }
+      return yearArray;
+    },
+
+    getIndicators() {
+      let indicArray = [];
+      for (let indicatorVal in this.indicators) {
+        indicArray.push([indicatorVal, this.indicators[indicatorVal].indicator_name]);
+      }
+      return indicArray;
+    },
+
+    getIndicatorName(indicatorId) {
+      return this.indicators[indicatorId].indicator_name;
+    },
+
+    reduceLegend(rangeNum) {
+      for (let indicatorVal in this.indicators) {
+        for (let yearVal in this.indicators[indicatorVal]) {
+          let year = this.indicators[indicatorVal][yearVal];
+          if ((typeof year.reduced === 'boolean') && (year.reduced === false)) {
```

Comment:
If you're checking the type perhaps it's better to rethink if that field should be split in 2 different ones.